### DISCUSSION TITLE
How to prevent Babel from renaming variable name "a" to "_a"

### DISCUSSION THREAD

**Discussion by mayewei** [babel/babel#12067] [Q&A] [answered]
How to prevent Babel from converting "a" to "_a"

**Comment by nicolo-ribaudo** [comment, marked as answer]
Variable names aren't semantically observable: even if Babel renames a variable, your program will work the same (ignoring `eval`, which isn't supported by Babel anyway).

That said, it usually renames a variable only when it's needed. Can you show an example of `a` becoming `_a`?

**Reply by mayewei** [reply]
I just want to escape JS of a project to Es5 version, but I use this variable name in HTML.

**Reply by nicolo-ribaudo** [reply]
Can you show a code example that is renamed by Babel?

**Reply by mayewei** [reply]
![image](https://user-images.githubusercontent.com/34051837/93451594-f5146d80-f909-11ea-88c7-189261e8f2e0.png)

**Reply by nicolo-ribaudo** [reply]
I tried copying parts of your code to our repl (https://babeljs.io/repl) but I coundn't reproduce it. Is it possible for you to create an example that I can try running? Are you only using Babel or also other tools?

Btw, if your goal is to use those variables in the HTML code I suggest using `window.tradeList` instead of `var tradeList`: by doing so no tool will renamed it, because it's a normal property on an object.

**Reply by mayewei** [reply]
```js
$(function () {
  login.init(); // $("#goByPassCodeA").hide();
});

var login = {
  init:function(){
  },
  login: function () {
  
    dataOP.post({
      url: dataOP.url.login.login,
      data: datas,
      success: function (data) {  
        
      },
      error: function (error) {
        login.toast(error.message);
        $('.load-wrapp').hide();
        $('.login-btn').show();
      }
    });
  }
};
```

Hello, you try this code, thank you

**Reply by nicolo-ribaudo** [reply]
Thanks! I can now explain _why_ that variable is renamed.

You are probably using `@babel/preset-env` (or, if you are using `@babel/preset-es2015`, you should replace it with `@babel/preset-env`).
`@babel/preset-env` includes a plugin called `@babel/plugin-transform-function-name`, which adds a name to functions inside of objects identical to the name of the object property. This means that, for example, this code:
```js
var obj = {
  fn: function() {}
}
```
becomes
```js
var obj = {
  fn: function fn() {}
}
```

Now, let's see what this plugin would do to your code:
```js
$(function () {
  login.init(); // $("#goByPassCodeA").hide();
});

var login = {
  init:function init(){
  },
  login: function login() {
  
    dataOP.post({
      url: dataOP.url.login.login,
      data: datas,
      success: function sucess(data) {  
        
      },
      error: function error(error) {
        login.toast(error.message);
        $('.load-wrapp').hide();
        $('.login-btn').show();
      }
    });
  }
};
```

By doing do, the `login` variable in `login.toast(error.message);` doesn't refer to the `var login` anymore, but to `function login() { ... }`. The only way that Babel has to refer to the correct variable is to rename it (to `_login`), so that it isn't shadowed anymore.

You can avoid this problem in two ways:
1. Disable `@babel/plugin-transform-function-name` in your config. It should look similar to this:
   ```json
   {
      "presets": [
         ["@babel/preset-env", {
            "exclude": ["transform-function-name"]
         }]
      ]
   }
   ```
2. Don't use the `login` variable inside the `login` function. You can, for example, introduce an `l` variable alias:
```js
$(function () {
  login.init(); // $("#goByPassCodeA").hide();
});

var login = {
  init:function(){
  },
  login: function () {
  
    dataOP.post({
      url: dataOP.url.login.login,
      data: datas,
      success: function (data) {  
        
      },
      error: function (error) {
        l.toast(error.message);
        $('.load-wrapp').hide();
        $('.login-btn').show();
      }
    });
  }
};
var l = login; // NOTE: Needed otherwise Babel renames login to _login.
```

**Reply by mayewei** [reply]
I tried your method and solved my problem. Thank you very much.
